presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipFunction.java

```diff
@@ -0,0 +1,129 @@
+    private static final String DESCRIPTION = "Aggregate elements of the arrays passed as arguments into rows.";
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ZipFunction.class, "zip", List.class, Block[].class);
+
+    static
+    {
+        ZIP_FUNCTIONS = new ZipFunction[MAX_ARITY - MIN_ARITY + 1];
+        for (int arity = MIN_ARITY; arity <= MAX_ARITY; ++arity) {
+            ZIP_FUNCTIONS[arity - MIN_ARITY] = new ZipFunction(arity);
+        }
+    }
+
+    private List<String> constraintNames;
+
+    private ZipFunction(int arity)
+    {
+        this(IntStream.rangeClosed(1, arity).mapToObj(s -> "T" + s).collect(toImmutableList()));
+    }
+
+    private ZipFunction(List<String> constraintNames)
+    {
+        super(FUNCTION_NAME,
+                constraintNames.stream().map(Signature::typeVariable).collect(toImmutableList()),
+                ImmutableList.of(),
+                "array(row(" + join(", ", constraintNames) + "))",
+                constraintNames.stream().map(name -> "array(" + name + ")").collect(toImmutableList()));
+        this.constraintNames = constraintNames;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        List<Type> constraintTypes = this.constraintNames.stream().map(constraintName -> boundVariables.getTypeVariable(constraintName)).collect(toImmutableList());
+        List<Boolean> nullableArguments = constraintTypes.stream().map(constraintType -> true).collect(toImmutableList());
+        List<Class<?>> javaArgumentTypes = constraintTypes.stream().map(constraintType -> Block.class).collect(toImmutableList());
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(constraintTypes).asVarargsCollector(Block[].class).asType(methodType(Block.class, javaArgumentTypes));
+        return new ScalarFunctionImplementation(false, nullableArguments, methodHandle, isDeterministic());
+    }
+
+    @UsedByGeneratedCode
+    public static Block zip(List<Type> types, Block... arrays)
+    {
+        int biggestCardinality = 0;
+        for (Block array : arrays) {
+            if (array != null) {
+                biggestCardinality = Math.max(biggestCardinality, array.getPositionCount());
+            }
+        }
+        RowType rowType = new RowType(types, Optional.empty());
+        BlockBuilder outputBuilder = rowType.createBlockBuilder(new BlockBuilderStatus(), biggestCardinality);
+        for (int outputPosition = 0; outputPosition < biggestCardinality; ++outputPosition) {
+            BlockBuilder rowBuilder = new InterleavedBlockBuilder(types, new BlockBuilderStatus(), 1);
```

Comment:
Instead of creating an `InterleavedBlockBuilder` you can do this more efficiently with `beginBlockEntry()` and `closeEntry()`. For example, take a look at how `MaxOrMinByStateSerializer.serialize` is implemented